### PR TITLE
Add cp and cp_rm transfer implementations

### PR DESCRIPTION
**Pull request by joshmoore**
As suggested by Niko on the forums, cp and cp_rm
implementations make local transfer of files possible
even across filesystem boundaries. The trade-off is
of course that in the case of a failed transfer there
will be increased space used which will need to be
manually cleaned up.

See: https://www.openmicroscopy.org/community/viewtopic.php?f=5&t=7699

/cc @ximenesuk

### CODE REVIEW

**Comment by joshmoore** [comment]
It might be worth teaching the transfer implementations to cleanup *targets* on failure as well to overcome the trade-off described in the commit message.

File: components/blitz/src/ome/formats/importer/cli/CommandLineImporter.java [hunk 1]
```diff
@@ -341,7 +341,8 @@ public static void advUsage() {
             + "        Server-side options:\t\n"
             + "          ln              \t# Use hard-link.\n"
             + "          ln_s            \t# Use soft-link.\n"
-            + "          ln_rm           \t# Caution! Hard-link followed by source deletion.\n\n"
```
Comment (ximenesuk) [inline components/blitz/src/ome/formats/importer/cli/CommandLineImporter.java]: Did you intend to remove this line?

File: components/blitz/src/ome/formats/importer/transfers/CopyFileTransfer.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (C) 2014 University of Dundee & Open Microscopy Environment.
```
Comment (mtbc) [inline components/blitz/src/ome/formats/importer/transfers/CopyFileTransfer.java]: 2015?

File: components/blitz/src/ome/formats/importer/cli/CommandLineImporter.java (hunk 1 above)
Comment (joshmoore) [inline components/blitz/src/ome/formats/importer/cli/CommandLineImporter.java]: No

File: components/blitz/src/ome/formats/importer/transfers/CopyFileTransfer.java (hunk 2 above)
Comment (joshmoore) [inline components/blitz/src/ome/formats/importer/transfers/CopyFileTransfer.java]: and yes

**Comment by joshmoore** [comment]
Cleaned up my early morning mistakes.

File: components/blitz/src/ome/formats/importer/transfers/CopyFileTransfer.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.List;
+
+/**
+ * Local-only file transfer mechanism which makes use of the plaform
+ * copy command.
+ *
+ * This is only useful where the commands "cp source target" (Unix) or
+ * "copy source target" (Windows) will work.
+ *
+ * @since 5.0.7
```
Comment (mtbc) [inline components/blitz/src/ome/formats/importer/transfers/CopyFileTransfer.java]: we're backporting?

Comment (joshmoore) [inline components/blitz/src/ome/formats/importer/transfers/CopyFileTransfer.java]: That's the intent.

**Comment by joshmoore** [comment]
Will also need documenting.

**Comment by ximenesuk** [comment]
The imports for both `cp` and `cp_rm` appear to work okay but the file size is incorrectly recorded in the DB, the size is the same as a link. The first of the three imports here was with `cp`, the second with `ln` and the third with `upload`
```
193 | user-0_2/2015-01/14/                       | 11-56-26.715                                                    |
 194 | user-0_2/2015-01/14/                       | 11-56-26.715.log                                                |    32033
 195 | user-0_2/2015-01/14/11-56-26.715/          | IAGFP-Noc01_R3D.dv                                              |       36
 196 | user-0_2/2015-01/14/11-56-26.715/          | IAGFP-Noc01_R3D.dv.log                                          |       36
 197 | user-0_2/2015-01/14/                       | 11-59-13.675                                                    |
 198 | user-0_2/2015-01/14/                       | 11-59-13.675.log                                                |    32077
 199 | user-0_2/2015-01/14/11-59-13.675/          | IAGFP-Noc01_R3D.dv                                              |       36
 200 | user-0_2/2015-01/14/11-59-13.675/          | IAGFP-Noc01_R3D.dv.log                                          |       36
 201 | user-0_2/2015-01/14/                       | 12-08-04.384                                                    |
 202 | user-0_2/2015-01/14/                       | 12-08-04.384.log                                                |    32116
 203 | user-0_2/2015-01/14/12-08-04.384/          | IAGFP-Noc01_R3D.dv                                              | 20447232
 204 | user-0_2/2015-01/14/12-08-04.384/          | IAGFP-Noc01_R3D.dv.log                                          |    27995
```

File: components/blitz/src/ome/formats/importer/cli/CommandLineImporter.java [hunk 4]
```diff
@@ -342,6 +342,8 @@ public static void advUsage() {
             + "          ln              \t# Use hard-link.\n"
             + "          ln_s            \t# Use soft-link.\n"
             + "          ln_rm           \t# Caution! Hard-link followed by source deletion.\n\n"
```
Comment (ximenesuk) [inline components/blitz/src/ome/formats/importer/cli/CommandLineImporter.java]: This and the next line need one '\n` removing. Only the last option needs the extra newline.

**Comment by joshmoore** [comment]
@ximenesuk : I'm not seeing any difference in behavior of `size` for imported images:
```
(v)# jamoore@joshuas-mbp:/opt/ome6$ dist/bin/omero hql "select o.name, o.size from OriginalFile o order by id desc" -q --limit=10
 # | Col1              | Col2
---+-------------------+------
 0 | to-be-copied.fake | 36
 1 | 14-07-44.422.log  | 9098
 2 | 14-07-44.422      | None
 3 | to-be-linked.fake | 36
 4 | 14-07-27.992.log  | 9088
 5 | 14-07-27.992      | None
 6 | to-be-linked.fake | 36
 7 | 14-02-51.632.log  | 9088
 8 | 14-02-51.632      | None
 9 | to-be-copied.fake | 36
(10 rows)
```
```
# jamoore@joshuas-mbp:/opt/ome7/components/tools/OmeroPy$ ls -ltra /tmp/ome6/ManagedRepository/root_0/2015-01/19/14-0*/*.fake
...
-rw-r--r--  3 jamoore  wheel  3662 Jan 19 14:07 /tmp/ome6/ManagedRepository/root_0/2015-01/19/14-02-51.632/to-be-linked.fake
-rw-r--r--  1 jamoore  wheel  3668 Jan 19 14:07 /tmp/ome6/ManagedRepository/root_0/2015-01/19/14-07-44.422/to-be-copied.fake
```

i.e. all are reported too small.

**Comment by ximenesuk** [comment]
@joshmoore which is the same as I am seeing, no? The file size that is reported correctly is the uploaded import.

**Comment by joshmoore** [comment]
For uploaded, yes. I guess, I'm pointing out that `ln` is already broken (i.e. not related to this PR):
```
(v)# jamoore@joshuas-mbp:/opt/ome6$ dist/bin/omero hql "select o.name, o.size from OriginalFile o where o.name like '%fake' order by id desc" -q --limit=10
 # | Col1                | Col2
---+---------------------+------
 0 | to-be-uploaded.fake | 3668
 1 | to-be-copied.fake   | 36
 2 | to-be-linked.fake   | 36
 3 | to-be-linked.fake   | 36
 4 | to-be-copied.fake   | 36
(5 rows)
```

**Comment by joshmoore** [comment]
@ximenesuk : pushed a fix for the next merge build. Note: this did apply to all transfers not just the new ones. New results:
```
(v)# jamoore@joshuas-mbp:/opt/ome6$ dist/bin/omero hql "select o.name, o.size from OriginalFile o where o.name like '%fake' order by id desc" -q --limit=10
 # | Col1                | Col2
---+---------------------+------
 0 | to-be-linked.fake   | 3662
 1 | to-be-uploaded.fake | 3668
 2 | to-be-copied.fake   | 3668

(v)# jamoore@joshuas-mbp:/opt/ome6$ ls -l /tmp/fakes/
total 24
-rw-r--r--  1 jamoore  wheel  3668 Jan 19 14:07 to-be-copied.fake
-rw-r--r--  4 jamoore  wheel  3662 Jan 19 14:07 to-be-linked.fake
-rw-r--r--  1 jamoore  wheel  3668 Jan 19 20:34 to-be-uploaded.fake
```

**Comment by mtbc** [comment]
Travis failure.

**Comment by joshmoore** [comment]
Romio/tests. Restarted.

**Comment by ximenesuk** [comment]
Sizes look okay now. I did notice that for `ln_s` the size in the database is now the size of the end-point file (rather than the size of the link) - is that the expected behaviour? If so how will that affect disk usage results on a system with high use of `ln_s`?

**Comment by joshmoore** [comment]
Pushed NPE fix for tomorrow.

**Comment by joshmoore** [comment]
Regarding the space issue, I assume this wasn't being seen previously because we were only getting the `36 B` of the UUID file? In Java 6, I don't know if we're going to have a good way to detect symlinks. (/cc @mtbc) It might be that trying to handle the file size update is too much. I'd still like to get `cp*` in to 5.0 for testing though.

**Comment by mtbc** [comment]
Once we're up to Java 7 we may well be able to find out through the `java.nio.file.attribute` stuff.

**Comment by ximenesuk** [comment]
This all looks good. Merge away.

**Comment by joshmoore** [comment]
Thanks, @ximenesuk. Merging.

On the topic of `DiskUsage`, come to think of it, `ln` files may pose a similar issue: i.e. even after you delete them, your server will not have more available space. I think this one is still a bit beyond us.

**Comment by joshmoore** [comment]
--rebased-to #3379

**Pull request merged**